### PR TITLE
RestWS-950-Encounter Search Handler should support a visits parameter

### PR DESCRIPTION
**Pull request by sadamhussain-m**
## Issue I worked on

see https://issues.openmrs.org/browse/RESTWS-

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ ] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [ ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [ ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java [hunk 1]
```diff
@@ -231,10 +231,10 @@ public PatientIdentifier save(PatientIdentifier delegate) {
 		}
 		
 		if (needToAdd) {
+			service().savePatientIdentifier(delegate);
 			delegate.getPatient().addIdentifier(delegate);
 		}
-		
-		service().savePatientIdentifier(delegate);
+
```
Comment (mseaton) [inline omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java:237]: This seems unrelated to this ticket, was this done to support a different issue?

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/CustomDatatypeResource1_9.java [hunk 2]
```diff
@@ -129,6 +130,9 @@ protected PageableResult doGetAll(RequestContext context) throws ResponseExcepti
 	}
 	
 	private List<CustomDatatypeRepresentation> getAllCustomDatatypes() {
+		if(!Context.isAuthenticated()) {
+			throw new APIAuthenticationException("User must be authenticated!");
+		}
```
Comment (mseaton) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/CustomDatatypeResource1_9.java:135]: This seems unrelated to this ticket, was this done to support a different issue?

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java [hunk 3]
```diff
@@ -189,4 +191,61 @@ public void shouldReturnTheAuditInfoForTheFullRepresentation() throws Exception
 		
 		assertNotNull(PropertyUtils.getProperty(result, "auditInfo"));
 	}
+
```
Comment (mseaton) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java:194]: This seems unrelated to this ticket, was this done to support a different issue?

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java [hunk 4]
```diff
@@ -58,7 +57,8 @@ public SearchConfig getSearchConfig() {
 	public PageableResult search(RequestContext context) throws ResponseException {
 		String patientUuid = context.getRequest().getParameter("patient");
 		String encounterTypeUuid = context.getRequest().getParameter("encounterType");
-		
+		String[] visitsUuid=context.getRequest().getParameter("visits").split(",");
```
Comment (mseaton) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:60]: The name of this parameter should be singular `visit`, not plural `visits`.

File: omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/CustomDatatypeResource1_9.java (hunk 2 above)
Comment (mogoodrich) [inline omod-1.9/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_9/CustomDatatypeResource1_9.java:135]: Same here, this looks unrelated, I think you are modifying another recent commit unintentionally?

**Review by mogoodrich — CHANGES_REQUESTED** [review]
Thanks @sadamhussain-m , generally looks good, but see my comments... there seem to be changes from other recent commits in your PR, and there is one place where I believe we need a null chek.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java (hunk 4 above)
Comment (mseaton) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:60]: Better to use existing method:  `String[] visitUuids = context.getRequest().getParameterValues("visit");`

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java [hunk 5]
```diff
@@ -69,13 +69,24 @@ public PageableResult search(RequestContext context) throws ResponseException {
...
 		if (patient != null) {
 			EncounterSearchCriteriaBuilder encounterSearchCriteriaBuilder = new EncounterSearchCriteriaBuilder()
 					.setPatient(patient).setFromDate(fromDate).setToDate(toDate).setIncludeVoided(false);
 			if (encounterType != null) {
 				encounterSearchCriteriaBuilder.setEncounterTypes(Arrays.asList(encounterType));
 			}
-			
+			if (!visits.isEmpty()){
+				encounterSearchCriteriaBuilder.setVisits(visits);
+			}
```
Comment (mseaton) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:89]:
```
			if (visitUuids != null && visitUuids.length > 0) {
				List<Visit> visits = new ArrayList<>();
				for (String visitUuid : visitUuids) {
					visits.add(Context.getVisitService().getVisitByUuid(visitUuid));
				}
				encounterSearchCriteriaBuilder.setVisits(visits);
			}
```

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java [hunk 6]
```diff
@@ -69,13 +69,24 @@ public PageableResult search(RequestContext context) throws ResponseException {
...
 				.getResourceBySupportedClass(EncounterType.class)).getByUniqueId(encounterTypeUuid);
+
+		Collection<Visit> visits = new ArrayList<>();
+		if(visitsUuid.length>0) {
+
+			for (String uuid : visitsUuid) {
+				Visit visit = ((VisitResource1_9) Context.getService(RestService.class).getResourceBySupportedClass(Visit.class)).getByUniqueId(uuid);
+				visits.add(visit);
+			}
+		}
```
Comment (mseaton) [inline omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:80]: Might as well wait to do all this once you are actually doing the query below, rather than pre-emptively, since patient is current required.  If we want to make patient no longer required, and query on visit only, then we need to make sure the patient null check does not wrap everything.  See suggested code below.

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java [hunk 7]
```diff
@@ -189,4 +191,61 @@ public void shouldReturnTheAuditInfoForTheFullRepresentation() throws Exception
...
+		req.setContent(json.getBytes());
+
+		SimpleObject updatedPatientIdentifier = deserialize(handle(req));
+		Object uuid = PropertyUtils.getProperty(updatedPatientIdentifier, "uuid");
+		Object identifierValue = PropertyUtils.getProperty(updatedPatientIdentifier, "identifier");
+
+		assertEquals(patientIdentifierUuidThatShouldNotChange, uuid);
+		assertEquals(patientIdentifierNewValue, identifierValue);
+		assertEquals(patientIdentifierNewValue, patientIdentifier.getIdentifier());
+	}
```
Comment (mseaton) [inline omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java:250]: We _do_ need to add a unit test for this feature, which should be placed in the `EncounterController2_0Test.java" class, which is where the existing tests of the Encounter 2.0 search functionality live.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java [hunk 8]
```diff
@@ -231,10 +231,10 @@ public PatientIdentifier save(PatientIdentifier delegate) {
 		}
 		
 		if (needToAdd) {
+			service().savePatientIdentifier(delegate);
 			delegate.getPatient().addIdentifier(delegate);
 		}
-		
-		service().savePatientIdentifier(delegate);
```
[the remark this answers is not in the document]
Comment (sadamhussain-m) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java:237]: I don't know how other commits are tagged up on this issue branch.Before I'm raising the PR,I tried to skip the other unrelated commits.But I can't find a way for it.Could you help me on this?.

File: omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java (hunk 1 above)
Comment (sadamhussain-m) [reply omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/PatientIdentifierResource1_8.java:237]: I don't know how the unrelated commits are tagged up on this issue branch.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java (hunk 4 above)
Comment (mogoodrich) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:60]: Sorry, @sadamhussain-m my bad, I think I said to use "visits", but let's go with "visit"

**Review by mogoodrich — CHANGES_REQUESTED** [review]
Looks like you were merging against the wrong branch, Sadam, but @ibacher  was able fix it, thanks @ibacher ... I am no longer seeing the changes to the other files.  If you can address the comments @mseaton  and I made we should be good to merge, thanks!

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java [hunk 9]
```diff
@@ -9,14 +9,12 @@
  */
 package org.openmrs.module.webservices.rest.web.v1_0.search.openmrs2_0;
 
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Date;
-import java.util.List;
+import java.util.*;
```
[the remark this answers is not in the document]
Comment (sadamhussain-m) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:12]: I'll rectify this.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java (hunk 4 above)
Comment (sadamhussain-m) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:60]: okay fine,I'll modify it.

[the remark this answers is not in the document]
Comment (sadamhussain-m) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:60]: sure,I'll add it up.

File: omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java (hunk 6 above)
Comment (sadamhussain-m) [reply omod-2.0/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs2_0/EncounterSearchHandler2_0.java:80]: Okay.I'll add up those validations for visits before the query itself.

File: omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java (hunk 7 above)
Comment (sadamhussain-m) [reply omod-1.9/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_9/PatientIdentifierController1_9Test.java]: can you give me a test data for the visits as input and the expected output for the encounters for a patient to check for it?

**Comment by mogoodrich** [comment]
@sadamhussain-m thanks so much for working on this!  I'm actually going to go ahead and merge this in, as we have need for it right now for a feature we are building.  I will handle the changes @mseaton  and I suggested and tag you on them, thanks!
